biogeochem/EDCanopyStructureMod.F90

```diff
@@ -349,905 +377,315 @@ subroutine canopy_structure( currentSite , bc_in )
+               if (abs(layer_co(ic)%pd_area-cohort%c_area) > nearzero) then
+                  layer_co(ic)%pd_area = layer_co(ic)%pd_area + &
+                       (excess_area/remainder_area) * &
+                       (cohort%c_area - layer_co(ic)%pd_area)
+               end if
+            end do
+         end if
+                  
+
+      else !comp_excl_exp < 0
+
+         ! ------------------------------------------------------------------
+         ! Rank Ordered Case
+         ! ------------------------------------------------------------------
+
+         sumpd_area = 0._r8
+         ic  = 1
+         do while( ic<=n_layer .and. (promdem_area-sumpd_area)>co_area_target_precision) 
+
+            cohort => layer_co(ic)%p
+
+            ! Determine if the next cohorts in
+            ! order have the same height
+
+            group_area = cohort%c_area
+            ic_n       = ic
+            check_next:do while(ic_n<n_layer)
+               if( abs(cohort%height-layer_co(ic_n+1)%p%height) > similar_height_tol ) then
+                  exit check_next
+               else
+                  ic_n = ic_n + 1
+                  group_area = group_area+layer_co(ic_n)%p%c_area
+               end if
+            end do check_next
+
+            remainder_area = min(promdem_area-sumpd_area,group_area)
+            do ic_nn = ic,ic_n
+               layer_co(ic_nn)%pd_area = remainder_area*layer_co(ic_nn)%p%c_area/group_area
+               sumpd_area = sumpd_area + layer_co(ic_nn)%pd_area
+            end do
+
+            ic = ic_n + 1
+
+         end do
+      end if comp_excl_type
+
+      ! Check to make sure the changes are within bounds
+      do ic = 1,n_layer
+         cohort => layer_co(ic)%p
+         if( ((layer_co(ic)%pd_area - cohort%c_area) > co_area_target_precision ) .or. &
+             (layer_co(ic)%pd_area < 0._r8) ) then
+            write(fates_log(),*) 'negative,or more area than the cohort has is being promoted/demoted'
+            write(fates_log(),*) 'change: ',layer_co(ic)%pd_area
+            write(fates_log(),*) 'existing area:',cohort%c_area
+            write(fates_log(),*) 'excess: ',layer_co(ic)%pd_area - cohort%c_area
+            call endrun(msg=errMsg(sourcefile, __LINE__))
+         end if
+      end do
+      
+      ! Part 3:
+      ! Apply the area changes by splitting the cohort and re-assigning
+      ! either all or part of it to a new layer
+
+      ic_loop0: do ic = 1,n_layer
+
+         cohort => layer_co(ic)%p
+
+         ! If the dem/prom area is the same area as the
+         !    cohort itself, move the whole thing
+         ! If the dem/prom area is less than the cohort area
+         !    and not trivialy small (larger than precision
+         !    check), then split it and move part of it
+         ! If the dem/prom area is less than zero or larger than
+         !    the cohort area within precision checks then FAIL
```

Comment:
Missing failure block?

Reply:
The failure check was provided in the checks before this logical block, so I updated text